### ISSUE TITLE
Issue with H.264

### ISSUE THREAD

**Issue by lminiero** [centricular/gstwebrtc-demos#129]
Hi,

I checked the older issues, but I couldn't find anything related to the issue I think I've identified, so I decided to open a new issue myself.

What I did was simple. Pretty much like it was done in #98 I modified the default `webrtc-sendrecv.c` demo to send H.264 instead of VP8. The simple diff is the following:

```
--- webrtc-sendrecv.c	2019-08-08 17:40:38.093263014 +0200
+++ webrtc-sendrecv-h264.c	2019-08-08 17:47:39.784094787 +0200
@@ -278,6 +278,7 @@
 #define STUN_SERVER " stun-server=stun://stun.l.google.com:19302 "
 #define RTP_CAPS_OPUS "application/x-rtp,media=audio,encoding-name=OPUS,payload="
 #define RTP_CAPS_VP8 "application/x-rtp,media=video,encoding-name=VP8,payload="
+#define RTP_CAPS_H264 "application/x-rtp,media=video,encoding-name=H264,payload="
 
 static void
 data_channel_on_error (GObject * dc, gpointer user_data)
@@ -335,8 +336,10 @@
 
   pipe1 =
       gst_parse_launch ("webrtcbin bundle-policy=max-bundle name=sendrecv " STUN_SERVER
-      "videotestsrc is-live=true pattern=ball ! videoconvert ! queue ! vp8enc deadline=1 ! rtpvp8pay ! "
-      "queue ! " RTP_CAPS_VP8 "96 ! sendrecv. "
+      //~ "videotestsrc is-live=true pattern=ball ! videoconvert ! queue ! vp8enc deadline=1 ! rtpvp8pay ! "
+      //~ "queue ! " RTP_CAPS_VP8 "96 ! sendrecv. "
+      "videotestsrc is-live=true pattern=ball ! videoconvert ! queue ! x264enc bitrate=500 speed-preset=veryfast tune=zerolatency ! video/x-h264,profile=baseline ! rtph264pay ! "
+      "queue ! " RTP_CAPS_H264 "96 ! sendrecv. "
       "audiotestsrc is-live=true wave=red-noise ! audioconvert ! audioresample ! queue ! opusenc ! rtpopuspay ! "
       "queue ! " RTP_CAPS_OPUS "97 ! sendrecv. ",
       &error);
```

The issue is that, when you try and do a call with a browser via `webrtc.nirbheek.in`, video from gstreamer works, but audio from gstreamer doesn't (no red noise). If you invert video and audio in the `gst_parse_launch()` call, so that the SDP contains an audio m-line first, then you get neither audio nor video. Audio and video coming from the browser always work, instead, as `autoaudiosink` and `autovideosink` always render media properly, which also means the PeerConnection is fine.

Everything works as expected when using VP8, no matter the order of audio and video in the SDP, which makes me suspect there's something in either `x264enc` or `rtph264pay` that may be confusing the pipeline, or make it get stuck somewhere. Or maybe I'm just crafting the H.264 pipeline part wrong?

Please let me know if there's anything else I can do or provide to help you investigate what may be the root cause of the issue, thanks!

**Comment by lminiero** [comment]
Sorry, forgot to mention... I'm using gstreamer 1.16 as available on Fedora 30.

**Comment by thaytan** [comment]
This is probably https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/issues/1028

**Comment by polhaghverdian** [comment]
Am trying with H264, the browser is not receiving the video.

`#define RTP_CAPS_OPUS "application/x-rtp,media=audio,encoding-name=OPUS,payload="
#define RTP_CAPS_H264 "application/x-rtp,media=video,encoding-name=H264,payload="
#define RTP_CAPS_VP8 "application/x-rtp,media=video,encoding-name=VP8,width=1280,height=720,framerate=60/1,payload="`

`		pipe1 =
			gst_parse_launch("webrtcbin bundle-policy=max-bundle name=sendrecv " STUN_SERVER
				"gdiscreencapsrc cursor=true ! videoconvert ! queue ! x264enc bitrate=500 speed-preset=veryfast tune=zerolatency ! video/x-h264,profile=baseline ! rtph264pay config-interval=1 ! "
				"queue ! " RTP_CAPS_H264 "96 ! sendrecv.  "
				"audiotestsrc is-live=true wave=red-noise ! audioconvert ! audioresample ! queue ! opusenc ! rtpopuspay ! "
				"queue ! " RTP_CAPS_OPUS "97 ! sendrecv. ",
				&error);`

**Comment by polhaghverdian** [comment]
I had this problem previously with Gstreamer **1.16**, but was fixed in release: **1.16.1**

**Comment by ystreet** [comment]
These examples have moved to https://gitlab.freedesktop.org/gstreamer/gst-examples/-/tree/master/webrtc and this repository will not be updated further.
